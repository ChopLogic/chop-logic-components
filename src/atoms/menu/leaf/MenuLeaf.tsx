import React from 'react';
import { ChopLogicIcon } from '@atoms';
import { ChopLogicMenuItem } from '@models';
import styles from './MenuLeaf.module.scss';

export const MenuLeaf = ({ item }: { item: ChopLogicMenuItem }): React.ReactElement => {
  const { icon, link, label, onFocus, onClick, onHover } = item;

  const leafContent = link ? (
    <a href={link} target='_blank' rel='noreferrer'>
      {label}
    </a>
  ) : (
    <span>{label}</span>
  );

  return (
    <li className={styles.menuLeaf} tabIndex={0} role='menuitem' onFocus={onFocus} onClick={onClick} onMouseOver={onHover}>
      <ChopLogicIcon name={icon} />
      {leafContent}
    </li>
  );
};
