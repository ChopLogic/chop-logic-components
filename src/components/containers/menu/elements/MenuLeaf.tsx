import React from 'react';

import { ChopLogicIcon } from '@/elements';
import { ChopLogicMenuItem, ChopLogicTheme } from '@/models';

import { StyledMenuLeaf } from '../Meny.styled';

type MenuLeafProps = {
  item: ChopLogicMenuItem;
  theme: ChopLogicTheme;
};

const MenuLeaf: React.FC<MenuLeafProps> = ({ theme, item }) => {
  const { icon, link, label, onFocus, onClick, onHover } = item;

  const leafContent = link ? (
    <a href={link} target='_blank' rel='noreferrer'>
      {label}
    </a>
  ) : (
    <span>{label}</span>
  );

  return (
    <StyledMenuLeaf tabIndex={0} role='menuitem' $theme={theme} onFocus={onFocus} onClick={onClick} onMouseOver={onHover}>
      <ChopLogicIcon name={icon} />
      {leafContent}
    </StyledMenuLeaf>
  );
};

export default MenuLeaf;
