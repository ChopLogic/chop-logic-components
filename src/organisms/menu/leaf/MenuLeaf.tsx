import { Icon } from '@atoms';
import type { MenuItem } from '@models';
import type { ReactElement } from 'react';

import styles from './MenuLeaf.module.scss';

export const MenuLeaf = ({ item }: { item: MenuItem }): ReactElement => {
  const { icon, link, label, onFocus, onClick, onHover } = item;

  const leafContent = link ? (
    <a href={link} target="_blank" rel="noreferrer">
      {label}
    </a>
  ) : (
    <span>{label}</span>
  );

  const handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    e.stopPropagation();
    if (e.key === 'Enter') {
      onClick?.();
    }
  };

  return (
    <li
      className={styles.menuLeaf}
      tabIndex={0}
      role="menuitem"
      onFocus={onFocus}
      onClick={onClick}
      onKeyDown={handleKeyDown}
      onMouseOver={onHover}
    >
      <Icon name={icon} hidden />
      {leafContent}
    </li>
  );
};
