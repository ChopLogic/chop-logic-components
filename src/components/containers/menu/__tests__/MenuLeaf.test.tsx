import { ChopLogicIconName } from '@enums';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import MenuLeaf from '../elements/MenuLeaf.tsx';

import { LIGHT_THEME } from '@css';

describe('MenuLeaf', () => {
  const testLabeledItem = {
    id: 'item-1',
    label: 'Go Home',
    icon: ChopLogicIconName.Home,
    onClick: vi.fn(),
  };

  const testLinkItem = {
    id: 'item-1',
    label: 'Go Home',
    icon: ChopLogicIconName.Home,
    link: 'https://example.com/',
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<MenuLeaf item={testLabeledItem} theme={LIGHT_THEME} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the link', () => {
    render(<MenuLeaf item={testLinkItem} theme={LIGHT_THEME} />);
    expect(screen.getByRole('link')).toBeInTheDocument();
  });

  it('should have accessible role', () => {
    render(<MenuLeaf item={testLinkItem} theme={LIGHT_THEME} />);
    expect(screen.getByRole('menuitem')).toBeInTheDocument();
  });

  it('should call onClick handler', async () => {
    render(<MenuLeaf item={testLabeledItem} theme={LIGHT_THEME} />);
    await userEvent.click(screen.getByText(testLabeledItem.label));
    expect(testLabeledItem.onClick).toHaveBeenCalledOnce();
  });
});
