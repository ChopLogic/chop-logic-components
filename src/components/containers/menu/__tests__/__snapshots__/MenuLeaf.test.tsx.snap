// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuLeaf > should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="sc-ivxoEo lcleEw"
    role="menuitem"
    tabindex="0"
  >
    <span
      aria-label="SVG Icon Fallback"
      style="font-size: 1rem;"
    >
      â¬œ
    </span>
    <span>
      Go Home
    </span>
  </li>
</DocumentFragment>
`;
